### PR TITLE
ON-38896 # update getElementSubmissionValue to treat propertyName and…

### PR DESCRIPTION
**Pull request by divporter**
… elementId separately

### CODE REVIEW

File: src/replaceCustomValues.ts [hunk 1]
```diff
@@ -221,22 +317,15 @@ export function getElementSubmissionValue({
         continue
       }
 
-      const result = getElementSubmissionValue({
+      const result = getElementSubmissionValueById({
         elementId,
-        propertyName,
         formElements: element.elements,
```
Comment (mymattcarroll) [inline src/replaceCustomValues.ts:322]: When iterating over pages, we need to ensure that the elements available are all elements across all pages/sections.

File: src/replaceCustomValues.ts [hunk 2]
```diff
@@ -221,22 +317,15 @@ export function getElementSubmissionValue({
...
-        formatNumber,
-        formatCurrency,
       })
       if (result) {
-        return result.value !== undefined || result.value !== null
-          ? result
-          : undefined
+        unknown = result.unknownValue
+        formElement = result.formElement
       }
```
Comment (mymattcarroll) [inline src/replaceCustomValues.ts:329]: can we return here to prevent further interations?

Comment (mymattcarroll) [reply src/replaceCustomValues.ts:329]: And then we dont need a return after the iterations?

File: src/replaceCustomValues.ts (hunk 1 above)
Comment (mymattcarroll) [reply src/replaceCustomValues.ts:322]: As discussed, I was confusing this logic with replacing values in context.

File: src/replaceCustomValues.ts (hunk 2 above)
Comment (divporter) [reply src/replaceCustomValues.ts:329]: nup

Comment (divporter) [reply src/replaceCustomValues.ts:329]: oh wait nevermind

File: src/replaceCustomValues.ts [hunk 3]
```diff
@@ -175,38 +175,134 @@ export function getElementSubmissionValue({
...
-    }
-    return 'name' in element && element.name === propertyName
-  }
-
   for (const element of formElements) {
-    if (matchFn(element)) {
+    if (elementId === element.id) {
       if ('name' in element) {
         unknown = submission[element.name]
         formElement = element
```
Comment (mymattcarroll) [inline src/replaceCustomValues.ts:308]: Same here, can we return early?

File: src/replaceCustomValues.ts [hunk 4]
```diff
@@ -175,68 +175,155 @@ export function getElementSubmissionValue({
...
+    (element) =>
+      element.type !== 'page' &&
+      element.type !== 'section' &&
+      element.name === propertyName,
+  )
+
+  const unknownValue = submission[propertyName]
+  if (unknownValue === undefined || unknownValue === null) {
+    return undefined
+  }
```
Comment (mymattcarroll) [inline src/replaceCustomValues.ts:279]:
```suggestion
  const unknownValue = submission[propertyName]
  if (unknownValue === undefined || unknownValue === null) {
    return undefined
  }
  
  const formElement = findFormElement(
    formElements,
    (element) =>
      element.type !== 'page' &&
      element.type !== 'section' &&
      element.name === propertyName,
  )
```

Can avoid iterating over elements if we dont need to.

File: src/replaceCustomValues.ts [hunk 5]
```diff
@@ -175,68 +175,155 @@ export function getElementSubmissionValue({
...
+  | { formElement: FormTypes.FormElement | undefined; unknownValue: unknown }
+  | undefined {
   // make sure submission is an object
   if (Object(submission) !== submission) {
     return undefined
   }
 
+  const flattenedElements = flattenFormElements(formElements)
+
   // initialise if propertyName provided and not elementId
```
Comment (mymattcarroll) [inline src/replaceCustomValues.ts:302]: Comment is outdated

File: src/replaceCustomValues.ts [hunk 6]
```diff
@@ -175,68 +175,155 @@ export function getElementSubmissionValue({
...
+function getElementSubmissionValueById({
+  elementId,
+  formElements,
+  submission,
+}: {
+  elementId: string
+  formElements: FormTypes.FormElement[]
+  submission: SubmissionTypes.S3SubmissionData['submission']
+}):
+  | { formElement: FormTypes.FormElement | undefined; unknownValue: unknown }
```
Comment (mymattcarroll) [inline src/replaceCustomValues.ts:293]:
```suggestion
  | { formElement: FormTypes.FormElement; unknownValue: unknown }
```

This should never return `undefined` for `formElement` now that we dont have to cater for a unknown element based on `propertyName`.

**Pull request merged**